File: image_processing_lib/src/image_adjust.c
/**
 * @file image_adjust.c
 * @brief Implementation of image adjustment operations
 */

#include "../include/image_adjust.h"
#include <string.h>

int adjust_brightness(const BMPImage *input, BMPImage *output, int factor) {
    int i;
    int temp;
    
    /* Copy header and properties */
    if (bmp_copy(input, output) != 0) {
        return -1;
    }
    
    /* Adjust brightness for each pixel */
    for (i = 0; i < input->dataSize; i++) {
        temp = input->data[i] + factor;
        
        /* Clamp to valid range [0, 255] */
        if (temp > 255) temp = 255;
        if (temp < 0) temp = 0;
        
        output->data[i] = (unsigned char)temp;
    }
    
    return 0;
}

int calculate_histogram(const BMPImage *input, float histogram[256]) {
    int x, y, i, j;
    long int ihist[256];
    long int sum;
    int imgRows = input->height;
    int imgCols = input->width;
    
    /* Check if image is 8-bit grayscale */
    if (input->bitDepth > 8) {
        printf("Error: Histogram requires 8-bit grayscale image\n");
        return -1;
    }
    
    /* Initialize histogram */
    for (i = 0; i < 256; i++) {
        ihist[i] = 0;
    }
    
    /* Count pixel intensities */
    sum = 0;
    for (y = 0; y < imgRows; y++) {
        for (x = 0; x < imgCols; x++) {
            j = input->data[x + y * imgCols];
            ihist[j]++;
            sum++;
        }
    }
    
    /* Normalize histogram */
    for (i = 0; i < 256; i++) {
        histogram[i] = (float)ihist[i] / (float)sum;
    }
    
    return 0;
}

int histogram_equalization(const BMPImage *input, BMPImage *output) {
    int x, y, i, j;
    int histeq[256];
    float hist[256];
    float sum;
    int imgRows = input->height;
    int imgCols = input->width;
    
    /* Check if image is 8-bit grayscale */
    if (input->bitDepth > 8) {
        printf("Error: Histogram equalization requires 8-bit grayscale image\n");
        return -1;
    }
    
    /* Calculate histogram */
    if (calculate_histogram(input, hist) != 0) {
        return -1;
    }
    
    /* Copy header and properties */
    if (bmp_copy(input, output) != 0) {
        return -1;
    }
    
    /* Calculate cumulative distribution function (CDF) */
    for (i = 0; i < 256; i++) {
        sum = 0.0;
        for (j = 0; j <= i; j++) {
            sum += hist[j];
        }
        histeq[i] = (int)(255 * sum + 0.5);
    }
    
    /* Apply histogram equalization */
    for (y = 0; y < imgRows; y++) {
        for (x = 0; x < imgCols; x++) {
            output->data[x + y * imgCols] = histeq[input->data[x + y * imgCols]];
        }
    }
    
    return 0;
}

int rotate_90_cw(const BMPImage *input, BMPImage *output) {
    int x, y;
    int imgRows = input->height;
    int imgCols = input->width;
    
    /* Copy header and update dimensions */
    memcpy(output->header, input->header, BMP_HEADER_SIZE);
    memcpy(output->colorTable, input->colorTable, BMP_COLOR_TABLE_SIZE);
    
    /* Swap width and height */
    output->width = imgRows;
    output->height = imgCols;
    output->bitDepth = input->bitDepth;
    output->dataSize = input->dataSize;
    
    /* Allocate memory */
    if (bmp_alloc_data(output, input->dataSize) != 0) {
        return -1;
    }
    
    /* Rotate pixels 90 degrees clockwise */
    for (y = 0; y < imgRows; y++) {
        for (x = 0; x < imgCols; x++) {
            output->data[(imgCols - 1 - x) * imgRows + y] = input->data[y * imgCols + x];
        }
    }
    
    return 0;
}

int rotate_180(const BMPImage *input, BMPImage *output) {
    int i;
    
    /* Copy header and properties */
    if (bmp_copy(input, output) != 0) {
        return -1;
    }
    
    /* Reverse pixel data */
    for (i = 0; i < input->dataSize; i++) {
        output->data[i] = input->data[input->dataSize - 1 - i];
    }
    
    return 0;
}

int rotate_90_ccw(const BMPImage *input, BMPImage *output) {
    int x, y;
    int imgRows = input->height;
    int imgCols = input->width;
    
    /* Copy header and update dimensions */
    memcpy(output->header, input->header, BMP_HEADER_SIZE);
    memcpy(output->colorTable, input->colorTable, BMP_COLOR_TABLE_SIZE);
    
    /* Swap width and height */
    output->width = imgRows;
    output->height = imgCols;
    output->bitDepth = input->bitDepth;
    output->dataSize = input->dataSize;
    
    /* Allocate memory */
    if (bmp_alloc_data(output, input->dataSize) != 0) {
        return -1;
    }
    
    /* Rotate pixels 90 degrees counter-clockwise */
    for (y = 0; y < imgRows; y++) {
        for (x = 0; x < imgCols; x++) {
            output->data[x * imgRows + (imgRows - 1 - y)] = input->data[y * imgCols + x];
        }
    }
    
    return 0;
}

File: src/image_processing/image_adjust.rs
use super::bmp_io::BmpImage;

// ---------------------------------------------------------------------------
// Brightness adjustment
// ---------------------------------------------------------------------------

/// Adjust brightness by adding `factor` to every pixel.
///
/// Positive values brighten, negative values darken.  Results are clamped to
/// [0, 255].
///
/// Translated from the C `ImageBrightness`, `ImageDarkness`, and
/// `ImageBrightnessCorrection` programs.
#[must_use]
pub fn adjust_brightness(input: &BmpImage, factor: i32) -> BmpImage {
    let output_data: Vec<u8> = input
        .data
        .iter()
        .map(|&px| (i32::from(px) + factor).clamp(0, 255) as u8)
        .collect();
    input.with_data(output_data)
}

// ---------------------------------------------------------------------------
// Histogram computation
// ---------------------------------------------------------------------------

/// Compute the normalised intensity histogram of a grayscale image.
///
/// Returns an array of 256 `f32` values where `hist[i]` is the fraction of
/// pixels with intensity `i`.
///
/// Translated from the C `ImgHistogram` function.
#[must_use]
pub fn compute_histogram(input: &BmpImage) -> [f32; 256] {
    let mut counts = [0u64; 256];
    let mut total: u64 = 0;

    for &px in &input.data {
        counts[px as usize] += 1;
        total += 1;
    }

    let mut hist = [0.0f32; 256];
    if total > 0 {
        for i in 0..256 {
            hist[i] = counts[i] as f32 / total as f32;
        }
    }
    hist
}

// ---------------------------------------------------------------------------
// Histogram equalization
// ---------------------------------------------------------------------------

/// Apply histogram equalization to a grayscale image for contrast enhancement.
///
/// Computes the CDF from the histogram and remaps pixel intensities.
///
/// Translated from the C `ImgHistogramEqualization` function.
#[must_use]
pub fn histogram_equalization(input: &BmpImage) -> BmpImage {
    let hist = compute_histogram(input);

    let mut histeq = [0u8; 256];
    for i in 0..256 {
        let mut cdf: f32 = 0.0;
        for j in 0..=i {
            cdf += hist[j];
        }
        histeq[i] = (255.0 * cdf + 0.5).clamp(0.0, 255.0) as u8;
    }

    let output_data: Vec<u8> = input
        .data
        .iter()
        .map(|&px| histeq[px as usize])
        .collect();
    input.with_data(output_data)
}

// ---------------------------------------------------------------------------
// Image rotation
// ---------------------------------------------------------------------------

/// Rotate a grayscale image 90° clockwise.
///
/// Translated from the C library `rotate_90_cw` function:
/// `output[(cols-1-x)*rows + y] = input[y*cols + x]`
#[must_use]
pub fn rotate_90_cw(input: &BmpImage) -> BmpImage {
    let rows = input.height as usize;
    let cols = input.width as usize;
    let mut output_data = vec![0u8; input.data.len()];

    for y in 0..rows {
        for x in 0..cols {
            let src = y * cols + x;
            let dst = (cols - 1 - x) * rows + y;
            if src < input.data.len() && dst < output_data.len() {
                output_data[dst] = input.data[src];
            }
        }
    }

    input.with_data(output_data)
}

/// Rotate a grayscale image 90° counter-clockwise.
///
/// Translated from the C library `rotate_90_ccw` function:
/// `output[x*rows + (rows-1-y)] = input[y*cols + x]`
#[must_use]
pub fn rotate_90_ccw(input: &BmpImage) -> BmpImage {
    let rows = input.height as usize;
    let cols = input.width as usize;
    let mut output_data = vec![0u8; input.data.len()];

    for y in 0..rows {
        for x in 0..cols {
            let src = y * cols + x;
            let dst = x * rows + (rows - 1 - y);
            if src < input.data.len() && dst < output_data.len() {
                output_data[dst] = input.data[src];
            }
        }
    }

    input.with_data(output_data)
}

/// Rotate a grayscale image 180°.
///
/// Translated from the C library `rotate_180` function which reverses
/// the entire pixel data array.
#[must_use]
pub fn rotate_180(input: &BmpImage) -> BmpImage {
    let len = input.data.len();
    let mut output_data = vec![0u8; len];

    for i in 0..len {
        output_data[i] = input.data[len - 1 - i];
    }

    input.with_data(output_data)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn brightness_increase_clamps_at_255() {
        let mut img = BmpImage::new();
        img.width = 2;
        img.height = 2;
        img.data = vec![200, 200, 200, 200];
        let out = adjust_brightness(&img, 100);
        assert!(out.data.iter().all(|&px| px == 255));
    }

    #[test]
    fn brightness_decrease_clamps_at_0() {
        let mut img = BmpImage::new();
        img.width = 2;
        img.height = 2;
        img.data = vec![30, 30, 30, 30];
        let out = adjust_brightness(&img, -50);
        assert!(out.data.iter().all(|&px| px == 0));
    }

    #[test]
    fn histogram_sums_to_one() {
        let mut img = BmpImage::new();
        img.width = 4;
        img.height = 4;
        img.data = vec![0, 64, 128, 192, 0, 64, 128, 192, 0, 64, 128, 192, 0, 64, 128, 192];
        let hist = compute_histogram(&img);
        let sum: f32 = hist.iter().sum();
        assert!((sum - 1.0).abs() < 0.001);
    }

    #[test]
    fn rotate_180_is_involution() {
        let mut img = BmpImage::new();
        img.width = 3;
        img.height = 3;
        img.data = vec![1, 2, 3, 4, 5, 6, 7, 8, 9];
        let rotated = rotate_180(&rotate_180(&img));
        assert_eq!(rotated.data, img.data);
    }
}
